bottlecap/src/lifecycle/invocation/triggers/kinesis_event.rs

```diff
@@ -0,0 +1,252 @@
+#![allow(clippy::module_name_repetitions)]
+use std::collections::HashMap;
+
+use datadog_trace_protobuf::pb::Span;
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use tracing::debug;
+
+use crate::lifecycle::invocation::triggers::FUNCTION_TRIGGER_EVENT_SOURCE_ARN_TAG;
+use crate::lifecycle::invocation::{
+    processor::MS_TO_NS,
+    triggers::{Trigger, FUNCTION_TRIGGER_EVENT_SOURCE_TAG},
+};
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct KinesisEvent {
+    #[serde(rename = "Records")]
+    pub records: Vec<KinesisEventRecord>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct KinesisEventRecord {
+    #[serde(rename = "eventID")]
+    pub event_id: String,
+    #[serde(rename = "eventName")]
+    pub event_name: String,
+    #[serde(rename = "eventSourceARN")]
+    pub event_source_arn: String,
+    #[serde(rename = "eventVersion")]
+    pub event_version: String,
+    pub kinesis: KinesisRecord,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct KinesisRecord {
+    #[serde(rename = "approximateArrivalTimestamp")]
+    pub approximate_arrival_timestamp: f64,
+    #[serde(rename = "partitionKey")]
+    pub partition_key: String,
+}
+
+impl Trigger for KinesisEventRecord {
+    fn new(payload: Value) -> Option<Self> {
+        let records = payload.get("Records").and_then(Value::as_array);
+        match records {
+            Some(records) => match serde_json::from_value::<KinesisEventRecord>(records[0].clone())
+            {
+                Ok(event) => Some(event),
+                Err(e) => {
+                    debug!("Failed to deserialize Kinesis Record: {e}");
+                    None
+                }
+            },
+            None => None,
+        }
+    }
+
+    fn is_match(payload: &Value) -> bool {
+        if let Some(first_record) = payload
+            .get("Records")
+            .and_then(Value::as_array)
+            .and_then(|r| r.first())
+            .take()
+        {
+            first_record.get("kinesis").is_some()
+        } else {
+            false
+        }
+    }
+
+    #[allow(clippy::cast_possible_truncation)]
+    fn enrich_span(&self, span: &mut Span) {
+        let event_source_arn = &self.event_source_arn;
+        let split_source_arn: Vec<&str> = event_source_arn.split('/').collect();
+        let parsed_stream_name = split_source_arn.last().unwrap_or(&"");
+        let parsed_shard_id = self.event_id.split(':').next().unwrap_or_default();
+        let service_name = "kinesis";
+        span.name = "aws.kinesis".to_string();
+        span.service = service_name.to_string();
+        span.start = (self.kinesis.approximate_arrival_timestamp * MS_TO_NS) as i64;
+        span.resource = (*parsed_stream_name).to_string();
```

Comment:
It was a ref of ref from the vec of &str
rust does a lot of deref under the hood automatically, I will remove the explicit deref. Can't use clone since it's &str -> String

Reply:
nevermind, not good
```
error: calling `to_string` on `&&str`
Error:   --> src/lifecycle/invocation/triggers/kinesis_event.rs:78:25
   |
78 |         span.resource = parsed_stream_name.to_string();
   |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try dereferencing the receiver: `(*parsed_stream_name).to_string()`
   ```